README.md
# sensu-dynamic-check-plugin

[![Sensu Bonsai Asset](https://img.shields.io/badge/Bonsai-Download%20Me-brightgreen.svg?colorB=89C967&logo=sensu)](https://bonsai.sensu.io/assets/betorvs/sensu-dynamic-check-mutator)
![Go Test](https://github.com/betorvs/sensu-dynamic-check-mutator/workflows/Go%20Test/badge.svg)
![goreleaser](https://github.com/betorvs/sensu-dynamic-check-mutator/workflows/goreleaser/badge.svg)

## Table of Contents
- [Overview](#overview)
- [Usage](#usage)
- [Configuration](#configuration)
    - [kubectl as asset](#kubectl-as-asset)
    - [http-nginx check example](#http-nginx)
  - [Json details](#json-details)
- [Asset registration](#asset-registration)
- [Mutator definition](#mutator-definition)
- [Installation from source](#installation-from-source)
- [Additional notes](#additional-notes)
- [Contributing](#contributing)

## Overview

The sensu-dynamic-check-plugin is a [Sensu Mutator][1] that parse labels from sensu events and creates a dynamic check in Sensu Backend and add event.check.annotation `io.sensu.remediation.config.actions` with these values then it can be scheduled by [sensu-remediation-handler][4]

## Usage

```bash
Sensu Dynamic Check Mutator creates sensu check based on template

Usage:
  sensu-dynamic-check-mutator [flags]
  sensu-dynamic-check-mutator [command]

Available Commands:
  help        Help about any command
  version     Print the version number of this plugin

Flags:
  -B, --api-backend-host string                  Sensu Go Backend API Host (e.g. 'sensu-backend.example.com') (default "127.0.0.1")
  -k, --api-backend-key string                   Sensu Go Backend API Key
  -P, --api-backend-pass string                  Sensu Go Backend API Password (default "P@ssw0rd!")
  -p, --api-backend-port int                     Sensu Go Backend API Port (e.g. 4242) (default 8080)
  -u, --api-backend-user string                  Sensu Go Backend API User (default "admin")
  -c, --check-config string                      Json template for Sensu Check
      --command-arguments-template string        Template for Sensu Check Command (default "{{ range $key, $value := . }} {{ $key }} {{ $value }}{{ end }}")
      --command-bool-arguments-template string   Template for Sensu Check Command (default "{{ range $value := . }} {{ $value }}{{ end }}")
      --command-handler string                   Handler used to post the result (default "default")
      --default-check-suffix-name string         Default suffix name for unpublished checks (default "dynamic")
  -h, --help                                     help for sensu-dynamic-check-mutator
  -i, --insecure-skip-verify                     skip TLS certificate verification (not recommended!)
      --remediation-event-annotation string      Add an annotation in dynamic check created like remediation-event-alias: event.Entity.Name/event.Check.Name (default "remediation-event-alias")
  -s, --secure                                   Use TLS connection to API
  -t, --trusted-ca-file string                   TLS CA certificate bundle in PEM format

Use "sensu-dynamic-check-mutator [command] --help" for more information about a command.

```

## Configuration

We add a json inside `--check-config`:

```json
[
  {
    "name": "describe-resource",
    "command": "${{assetPath \"kubectl\"}}/kubernetes/client/bin/kubectl describe",
    "bool_args": [
      "--no-headers"
    ],
    "arguments": ["daemonset","deployment","pod","statefulset"],
    "options": {
        "--namespace": "namespace"
    },
    "match_labels": {
        "sensu-alertmanager-events": "owner"
    },
    "exclude_labels": [
      {
        "alertname": "TargetDown"
      },
      {
        "alertname": "KubeVersionMismatch"
      }
    ],
    "sensu_assets": [
        "kubectl"
    ],
    "occurrences": [1],
    "severities": [2],
  },
  {
    "name": "systemctl-status",
    "command": "sudo systemctl",
    "options": {
        "status": "application"
    },
    "match_labels": {
        "systemd": "true"
    },
    "occurrences": [1]
  },
  {
    "name": "systemctl-restart",
    "command": "sudo systemctl",
    "options": {
        "restart": "application"
    },
    "match_labels": {
        "systemd": "true"
    },
    "occurrences": [3]
  }
]
```

In this example, to change the event, this mutator need to find a label called `namespace`, and need to find at least one of the arguments array, like label `deployment`. Then it will create a check.command: `${{assetPath "kubectl"}}/kubernetes/client/bin/kubectl describe --namespace default deployment nginx`.

And it will create one annotation like:

```
"io.sensu.remediation.config.actions": "[{\"request\":\"KubeDeploymentReplicasMismatch-default-nginx-describe-resource-dynamic\",\"occurrences\":[1],\"severities\":[2],\"subscriptions\":[\"entity:k8s.dev.local\"]}]"
```

And one check called: `KubeDeploymentReplicasMismatch-default-nginx-describe-resource-dynamic` with command `${{assetPath \"kubectl\"}}/kubernetes/client/bin/kubectl describe --namespace default daemonset nginx`.


In `systemctl-status` and `systemctl-restart` if this mutator found two labels `systemd:true` and `application:nginx` as example for a check called `http-nginx`, it will create two checks in Sensu Backend called `http-nginx-systemctl-status-dynamic` and `http-nginx-systemctl-restart-dynamic` both running a `sudo systemctl [status|restart] nginx` command and it will add the following annotation: 
```yml
"io.sensu.remediation.config.actions": "[{\"request\":\"http-nginx-systemctl-status-dynamic\",\"occurrences\":[1],\"severities\":[2],\"subscriptions\":[\"entity:systemd-ubuntu\"]},{\"request\":\"http-nginx-systemctl-restart-dynamic\",\"occurrences\":[3],\"severities\":[2],\"subscriptions\":[\"entity:systemd-ubuntu\"]}]"
```
   
#### kubectl as asset

In these example we use one event imported by [sensu-alertmanager-events][5] and we installed kubectl using assets.

```yml
type: Asset
api_version: core/v2
metadata:
  name: kubectl
  namespace: default
spec:
  sha512:  081472833601aa4fa78e79239f67833aa4efcb4efe714426cd01d4ddf6f36fbf304ef7e1f5373bff0fdff44a845f7560165c093c108bd359b5ab4189f36b1f2f
  url: https://dl.k8s.io/v1.20.0/kubernetes-client-linux-amd64.tar.gz
```

#### http-nginx 

```yml
type: Check
api_version: core/v2
metadata:
  name: http-nginx
  namespace: default
spec:
  command: check-http.rb -u http://127.0.0.1 -t 5
  handlers:
  - default
  - remediation
  interval: 60
  publish: true
  runtime_assets:
  - sensu-ruby-runtime
  - sensu-plugins-http
  subscriptions:
  - ubuntu
```


### Json details

| Field | What it does | Example |
| ----- | ------------ | ------- |
| bool_args | add flags without any argument. Always include any configured flags | `-k` |
| arguments | add label.key label.value inside command. Should match at least one. If not, will return event without any change | `deployment ingress-nginx` |
| options | should match all configured to change the event. Use it when you need to use a different flag but with some content from a label |To use a label.value in the flag `--namespace`, use it: `{"--namespace": "namespace"}`
| match_labels | If found these label.key=label.value it will change the event | - |
| exclude_labels | Use this array to exclude some label.key=label.value that doesnt match with your dynamic check | - |  
| occurrences | same occurrences field in [sensu-remediation-handler][4] | default: `[]int{1}` |
| severities | same severities field in [sensu-remediation-handler][4] | default: `[]int{2}` |
| publish | bool field. If it is enabled it will not send any information to sensu-remediation-handler | default: `false` |
| interval | integer field | default: `10` |
| subscription | string field used to overwrite subscription used in check definition created by sensu-dynamic-check-mutator| default: `""` |
| name_suffix | string field append in check name a label.value | default: `""` | 
| proxy_entity_id | string field used in check.proxy_entity_id based on label.value | default: `""` | 
| sensu_handlers | []string used to send handler with dynamic check created | default: `[]string{"default"}` |




### Asset registration

[Sensu Assets][2] are the best way to make use of this plugin. If you're not using an asset, please
consider doing so! If you're using sensuctl 5.13 with Sensu Backend 5.13 or later, you can use the
following command to add the asset:

```
sensuctl asset add betorvs/sensu-dynamic-check-mutator
```

If you're using an earlier version of sensuctl, you can find the asset on the [Bonsai Asset Index][https://bonsai.sensu.io/assets/betorvs/sensu-dynamic-check-mutator].

### Mutator definition

Maybe is important to add authetication configs `-u dynamic -P ${MUTATOR_PASS} -B sensu-api.k8s.infra.ppro.com -s -t /$PATH_TO_CERTIFICATE/ca.pem`

```yml
---
type: Mutator
api_version: core/v2
metadata:
  name: sensu-dynamic-check-mutator
  namespace: default
spec:
  command: >-
    sensu-dynamic-check-mutator -c "[{\"name\":\"describe-resource\",\"command\":\"\${{assetPath \\\"kubectl\\\"}}/kubernetes/client/bin/kubectl describe\",\"bool_args\":[\"--no-headers\"],\"arguments\":[\"daemonset\",\"deployment\",\"pod\",\"statefulset\",\"node\"],\"options\":{\"--namespace\":\"namespace\"},\"match_labels\":{\"sensu-alertmanager-events\":\"owner\"},\"exclude_labels\":[{\"alertname\":\"TargetDown\"},{"alertname": "KubeVersionMismatch"}],\"sensu_assets\":[\"kubectl\"]}]"
  runtime_assets:
  - betorvs/sensu-dynamic-check-mutator
```

## Installation from source

The preferred way of installing and deploying this plugin is to use it as an Asset. If you would
like to compile and install the plugin from source or contribute to it, download the latest version
or create an executable script from this source.

From the local path of the sensu-dynamic-check-mutator repository:

```
go build
```

## Additional notes

## Contributing

For more information about contributing to this plugin, see [Contributing][3].

[1]: https://docs.sensu.io/sensu-go/latest/reference/mutators/
[2]: https://docs.sensu.io/sensu-go/latest/reference/assets/
[3]: https://github.com/sensu/sensu-go/blob/master/CONTRIBUTING.md
[4]: https://github.com/sensu/sensu-remediation-handler
[5]: https://github.com/betorvs/sensu-alertmanager-events



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic
Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased

## [0.0.1] - 2020-01-01

### Added
- Initial release


main_test.go
package main

import (
	"testing"

	v2 "github.com/sensu/sensu-go/api/core/v2"
	"github.com/stretchr/testify/assert"
)

func TestCheckArgs(t *testing.T) {
	assert := assert.New(t)
	event := v2.FixtureEvent("entity1", "check1")
	err := checkArgs(event)
	assert.Error(err)
	event2 := v2.FixtureEvent("entity2", "check2")
	mutatorConfig.CheckConfig = "[{\"name\":\"describe-resource\",\"command\":\"${{assetPath \\\"kubectl\\\"}}/kubernetes/client/bin/kubectl describe\",\"bool_args\":[\"--no-headers\"],\"arguments\":[\"daemonset\",\"deployment\",\"pod\",\"statefulset\",\"node\"],\"options\":{\"--namespace\":\"namespace\"},\"match_labels\":{\"sensu-alertmanager-events\":\"owner\"},\"sensu_assets\":[\"kubectl\"]}]"
	err2 := checkArgs(event2)
	assert.NoError(err2)
}

func TestExtractLabels(t *testing.T) {
	event1 := v2.FixtureEvent("entity1", "check1")
	event1.Labels["test1"] = "value1"
	value1, result1 := extractLabels(event1, "test1")
	assert.Contains(t, value1, "value1")
	assert.True(t, result1)
	event2 := v2.FixtureEvent("entity2", "check2")
	_, result2 := extractLabels(event2, "test2")
	assert.False(t, result2)
}

func TestSearchLabels(t *testing.T) {
	event1 := v2.FixtureEvent("entity1", "check1")
	event1.Labels["testa"] = "valuea"
	event1.Labels["testb"] = "valueb"
	event1.Labels["testc"] = "valuec"
	labels := make(map[string]string)
	res1 := searchLabels(event1, labels)
	assert.False(t, res1)

	labels["testa"] = "valuea"
	labels["testc"] = "valuec"
	res2 := searchLabels(event1, labels)
	assert.True(t, res2)

	excludeLabels := make(map[string]string)
	excludeLabels["testc"] = "valuec"
	res3 := searchLabels(event1, excludeLabels)
	assert.True(t, res3)
}

func TestParseCommandOptions(t *testing.T) {
	test1 := make(map[string]string)
	test1["foo"] = "bar"
	test1["test1"] = "test-1"
	mutatorConfig.CommandArgumentsTemplate = "{{ range $key, $value := . }} {{ $key }} {{ $value }}{{ end }}"
	result1 := parseCommandOptions(test1)
	assert.Contains(t, result1, "test1 test-1")
}

func TestParseCommandBoolFlags(t *testing.T) {
	test1 := []string{"-A", "-k"}
	mutatorConfig.CommandBoolArgumentsTemplate = "{{ range $value := . }} {{ $value }}{{ end }}"
	result1 := parseCommandBoolFlags(test1)
	assert.Contains(t, result1, "-A")
}


main.go
package main

import (
	"bytes"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"strings"

	"github.com/sensu-community/sensu-plugin-sdk/sensu"
	"github.com/sensu-community/sensu-plugin-sdk/templates"
	v2 "github.com/sensu/sensu-go/api/core/v2"
	"github.com/sensu/sensu-go/types"
)

// CheckTemplate struct
type CheckTemplate struct {
	Name            string              `json:"name"`
	Command         string              `json:"command"`
	Arguments       []string            `json:"arguments"`
	Options         map[string]string   `json:"options"`
	BoolOptions     []string            `json:"bool_options"`
	MatchLabels     map[string]string   `json:"match_labels"`
	ExcludeLabels   []map[string]string `json:"exclude_labels"`
	SensuAssets     []string            `json:"sensu_assets"`
	Occurrences     []int               `json:"occurrences"`
	Severities      []int               `json:"severities"`
	Publish         bool                `json:"publish"`
	Interval        int                 `json:"interval"`
	Subscription    string              `json:"subscription"`
	NameSuffixLabel string              `json:"name_suffix"`
	ProxyEntityID   string              `json:"proxy_entity_id"`
	SensuHandlers   []string            `json:"sensu_handlers"`
}

// Auth represents the authentication info
type Auth struct {
	AccessToken  string `json:"access_token"`
	RefreshToken string `json:"refresh_token"`
	ExpiresAt    int64  `json:"expires_at"`
}

// RemediationConfig from sensu-remediation-handler
type RemediationConfig struct {
	Request       string   `json:"request"`
	Occurrences   []int    `json:"occurrences"`
	Severities    []int    `json:"severities"`
	Subscriptions []string `json:"subscriptions"`
}

// Config represents the mutator plugin config.
type Config struct {
	sensu.PluginConfig
	CheckConfig                  string
	CommandArgumentsTemplate     string
	CommandBoolArgumentsTemplate string
	CommandHandler               string
	APIBackendPass               string
	APIBackendUser               string
	APIBackendKey                string
	APIBackendHost               string
	APIBackendPort               int
	Secure                       bool
	TrustedCAFile                string
	InsecureSkipVerify           bool
	Protocol                     string
	DefaultCheckSuffixName       string
	RemediationEventAnnotation   string
}

var (
	tlsConfig tls.Config

	mutatorConfig = Config{
		PluginConfig: sensu.PluginConfig{
			Name:     "sensu-dynamic-check-mutator",
			Short:    "Sensu Dynamic Check Mutator creates sensu check based on template",
			Keyspace: "sensu.io/plugins/sensu-dynamic-check-mutator/config",
		},
	}

	options = []*sensu.PluginConfigOption{
		{
			Path:      "check-config",
			Env:       "",
			Argument:  "check-config",
			Shorthand: "c",
			Default:   "",
			Usage:     "Json template for Sensu Check",
			Value:     &mutatorConfig.CheckConfig,
		},
		{
			Path:      "command-arguments-template",
			Env:       "",
			Argument:  "command-arguments-template",
			Shorthand: "",
			Default:   "{{ range $key, $value := . }} {{ $key }} {{ $value }}{{ end }}",
			Usage:     "Template for Sensu Check Command",
			Value:     &mutatorConfig.CommandArgumentsTemplate,
		},
		{
			Path:      "command-bool-arguments-template",
			Env:       "",
			Argument:  "command-bool-arguments-template",
			Shorthand: "",
			Default:   "{{ range $value := . }} {{ $value }}{{ end }}",
			Usage:     "Template for Sensu Check Command",
			Value:     &mutatorConfig.CommandBoolArgumentsTemplate,
		},
		{
			Path:      "command-handler",
			Env:       "",
			Argument:  "command-handler",
			Shorthand: "",
			Default:   "default",
			Usage:     "Handler used to post the result",
			Value:     &mutatorConfig.CommandHandler,
		},
		{
			Path:      "api-backend-user",
			Env:       "SENSU_API_USER",
			Argument:  "api-backend-user",
			Shorthand: "u",
			Default:   "admin",
			Usage:     "Sensu Go Backend API User",
			Value:     &mutatorConfig.APIBackendUser,
		},
		{
			Path:      "api-backend-pass",
			Env:       "SENSU_API_PASSWORD",
			Argument:  "api-backend-pass",
			Shorthand: "P",
			Default:   "P@ssw0rd!",
			Usage:     "Sensu Go Backend API Password",
			Value:     &mutatorConfig.APIBackendPass,
		},
		{
			Path:      "api-backend-key",
			Env:       "SENSU_API_KEY",
			Argument:  "api-backend-key",
			Shorthand: "k",
			Default:   "",
			Usage:     "Sensu Go Backend API Key",
			Value:     &mutatorConfig.APIBackendKey,
		},
		{
			Path:      "api-backend-host",
			Env:       "",
			Argument:  "api-backend-host",
			Shorthand: "B",
			Default:   "127.0.0.1",
			Usage:     "Sensu Go Backend API Host (e.g. 'sensu-backend.example.com')",
			Value:     &mutatorConfig.APIBackendHost,
		},
		{
			Path:      "api-backend-port",
			Env:       "",
			Argument:  "api-backend-port",
			Shorthand: "p",
			Default:   8080,
			Usage:     "Sensu Go Backend API Port (e.g. 4242)",
			Value:     &mutatorConfig.APIBackendPort,
		},
		{
			Path:      "secure",
			Env:       "",
			Argument:  "secure",
			Shorthand: "s",
			Default:   false,
			Usage:     "Use TLS connection to API",
			Value:     &mutatorConfig.Secure,
		},
		{
			Path:      "insecure-skip-verify",
			Env:       "",
			Argument:  "insecure-skip-verify",
			Shorthand: "i",
			Default:   false,
			Usage:     "skip TLS certificate verification (not recommended!)",
			Value:     &mutatorConfig.InsecureSkipVerify,
		},
		{
			Path:      "trusted-ca-file",
			Env:       "",
			Argument:  "trusted-ca-file",
			Shorthand: "t",
			Default:   "",
			Usage:     "TLS CA certificate bundle in PEM format",
			Value:     &mutatorConfig.TrustedCAFile,
		},
		{
			Path:      "default-check-suffic-name",
			Env:       "",
			Argument:  "default-check-suffix-name",
			Shorthand: "",
			Default:   "dynamic",
			Usage:     "Default suffix name for unpublished checks",
			Value:     &mutatorConfig.DefaultCheckSuffixName,
		},
		{
			Path:      "remediation-event-annotation",
			Env:       "",
			Argument:  "remediation-event-annotation",
			Shorthand: "",
			Default:   "remediation-event-alias",
			Usage:     "Add an annotation in dynamic check created like remediation-event-alias: event.Entity.Name/event.Check.Name",
			Value:     &mutatorConfig.RemediationEventAnnotation,
		},
	}
)

func main() {
	mutator := sensu.NewGoMutator(&mutatorConfig.PluginConfig, options, checkArgs, executeMutator)
	mutator.Execute()
}

func checkArgs(_ *types.Event) error {
	if len(mutatorConfig.CheckConfig) == 0 {
		return fmt.Errorf("--check-config is required")
	}
	// For Sensu Backend Connections
	if mutatorConfig.Secure {
		mutatorConfig.Protocol = "https"
	} else {
		mutatorConfig.Protocol = "http"
	}
	if len(mutatorConfig.TrustedCAFile) > 0 {
		caCertPool, err := v2.LoadCACerts(mutatorConfig.TrustedCAFile)
		if err != nil {
			return fmt.Errorf("Error loading specified CA file")
		}
		tlsConfig.RootCAs = caCertPool
	}
	tlsConfig.InsecureSkipVerify = mutatorConfig.InsecureSkipVerify

	tlsConfig.CipherSuites = v2.DefaultCipherSuites

	return nil
}

func executeMutator(event *types.Event) (*types.Event, error) {
	// log.Println("executing mutator with --check-config", mutatorConfig.CheckConfig)
	checkTemplate := []CheckTemplate{}
	err := json.Unmarshal([]byte(mutatorConfig.CheckConfig), &checkTemplate)
	if err != nil {
		return event, err
	}
	remediations := []RemediationConfig{}
	for _, v := range checkTemplate {
		for _, e := range v.ExcludeLabels {
			if searchLabels(event, e) {
				return event, nil
			}
		}
		if searchLabels(event, v.MatchLabels) {
			// fmt.Printf("Check Name: %s\n", v.Name)
			command := v.Command
			var flags, args, boolFlags string
			if v.Options != nil {
				tempArgs := make(map[string]string)
				count := 0
				for key, value := range v.Options {
					temp, valid := extractLabels(event, value)
					if valid {
						tempArgs[key] = temp
						count++
					}
				}
				if len(v.Options) != count {
					return event, nil
				}
				flags = parseCommandOptions(tempArgs)
			}
			if len(v.Arguments) != 0 {
				tempArgs := make(map[string]string)
				count := 0
				for _, value := range v.Arguments {
					temp, valid := extractLabels(event, value)
					if valid {
						tempArgs[value] = temp
						count++
					}
				}
				if count < 1 {
					return event, nil
				}
				args = parseCommandOptions(tempArgs)
			}

			if len(v.BoolOptions) != 0 {
				boolFlags = parseCommandBoolFlags(v.BoolOptions)
			}

			if flags != "" {
				command += flags
			}
			if args != "" {
				command += args
			}
			if boolFlags != "" {
				command += boolFlags
			}
			tempName := fmt.Sprintf("%s-%s-%s", event.Check.Name, v.Name, mutatorConfig.DefaultCheckSuffixName)
			if v.NameSuffixLabel != "" {
				temp, valid := extractLabels(event, v.NameSuffixLabel)
				if valid {
					tempName = fmt.Sprintf("%s-%s-%s-%s", event.Check.Name, v.Name, temp, mutatorConfig.DefaultCheckSuffixName)
				}
			}
			var autherr error
			auth := Auth{}
			if len(mutatorConfig.APIBackendKey) == 0 {
				auth, autherr = authenticate()

				if autherr != nil {
					return event, autherr
				}
			}
			entity := fmt.Sprintf("entity:%s", event.Entity.ObjectMeta.Name)
			assets := []string{}
			if len(v.SensuAssets) != 0 {
				assets = v.SensuAssets
			}
			publish := false
			interval := 10
			if v.Publish {
				publish = v.Publish
			}
			if v.Interval != 0 {
				interval = v.Interval
			}
			var subscription string
			if v.Subscription != "" {
				subscription = v.Subscription
			}
			var proxyEntity string
			if v.ProxyEntityID != "" {
				temp, valid := extractLabels(event, v.ProxyEntityID)
				if valid {
					proxyEntity = temp
				}
			}
			handler := []string{"default"}
			if len(v.SensuHandlers) != 0 {
				handler = v.SensuHandlers
			}
			alias := fmt.Sprintf("%s/%s", event.Entity.Name, event.Check.Name)
			err := postCheck(auth, tempName, command, event.Namespace, entity, subscription, proxyEntity, alias, handler, assets, publish, interval)
			if err != nil {
				return event, err
			}
			if !publish {
				occurrence := []int{1}
				severity := []int{2}
				if len(v.Occurrences) != 0 {
					occurrence = v.Occurrences
				}
				if len(v.Severities) != 0 {
					severity = v.Severities
				}
				remediation := RemediationConfig{
					Request:       tempName,
					Occurrences:   occurrence,
					Severities:    severity,
					Subscriptions: []string{entity},
				}
				remediations = append(remediations, remediation)
			}
		}

	}

	s, _ := json.Marshal(remediations)
	annotations := make(map[string]string)
	annotations["io.sensu.remediation.config.actions"] = string(s)
	// copy all annotations from event.check
	if event.Check.Annotations != nil {
		for k, v := range event.Check.Annotations {
			annotations[k] = v
		}
	}
	// add new annotations map with grafana URLs
	event.Check.Annotations = annotations

	return event, nil
}

func parseCommandOptions(arguments map[string]string) (title string) {
	title, err := templates.EvalTemplate("title", mutatorConfig.CommandArgumentsTemplate, arguments)
	if err != nil {
		return ""
	}
	return title
}

func parseCommandBoolFlags(arguments []string) (title string) {
	title, err := templates.EvalTemplate("title", mutatorConfig.CommandBoolArgumentsTemplate, arguments)
	if err != nil {
		return ""
	}
	return title
}

func extractLabels(event *types.Event, label string) (string, bool) {
	labelFound := ""
	if event.Labels != nil {
		for k, v := range event.Labels {
			if k == label {
				labelFound = v
			}
		}
	}
	if event.Entity.Labels != nil {
		for k, v := range event.Entity.Labels {
			if k == label {
				labelFound = v
			}
		}
	}
	if event.Check.Labels != nil {
		for k, v := range event.Check.Labels {
			if k == label {
				labelFound = v
			}
		}
	}
	if labelFound == "" {
		return labelFound, false
	}
	return labelFound, true
}

func searchLabels(event *types.Event, labels map[string]string) bool {
	if len(labels) == 0 {
		return false
	}
	count := 0
	for key, value := range labels {
		if event.Labels != nil {
			for k, v := range event.Labels {
				if k == key && v == value {
					count++
				}
			}
		}
		if event.Entity.Labels != nil {
			for k, v := range event.Entity.Labels {
				if k == key && v == value {
					count++
				}
			}
		}
		if event.Check.Labels != nil {
			for k, v := range event.Check.Labels {
				if k == key && v == value {
					count++
				}
			}
		}
		if count == len(labels) {
			return true
		}
	}

	return false
}

// authenticate funcion to wotk with api-backend-* flags
func authenticate() (Auth, error) {
	var auth Auth
	client := http.DefaultClient
	client.Transport = http.DefaultTransport

	if mutatorConfig.Secure {
		client.Transport.(*http.Transport).TLSClientConfig = &tlsConfig
	}

	req, err := http.NewRequest(
		"GET",
		fmt.Sprintf("%s://%s:%d/auth", mutatorConfig.Protocol, mutatorConfig.APIBackendHost, mutatorConfig.APIBackendPort),
		nil,
	)
	if err != nil {
		return auth, fmt.Errorf("error generating auth request: %v", err)
	}

	req.SetBasicAuth(mutatorConfig.APIBackendUser, mutatorConfig.APIBackendPass)

	resp, err := client.Do(req)
	if err != nil {
		return auth, fmt.Errorf("error executing auth request: %v", err)
	}

	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return auth, fmt.Errorf("error reading auth response: %v", err)
	}

	if strings.HasPrefix(string(body), "Unauthorized") {
		return auth, fmt.Errorf("authorization failed for user %s", mutatorConfig.APIBackendUser)
	}

	err = json.NewDecoder(bytes.NewReader(body)).Decode(&auth)

	if err != nil {
		trim := 64
		return auth, fmt.Errorf("error decoding auth response: %v\nFirst %d bytes of response: %s", err, trim, trimBody(body, trim))
	}

	return auth, err
}

// post check to sensu-backend-api
func postCheck(auth Auth, name, command, namespace, entity, subscription, proxyEntity, alias string, handlers, assets []string, publish bool, interval int) error {
	client := http.DefaultClient
	client.Transport = http.DefaultTransport
	// /api/core/v2/namespaces/NAMESPACE/checks/:check_name and PUT
	url := fmt.Sprintf("%s://%s:%d/api/core/v2/namespaces/%s/checks/%s", mutatorConfig.Protocol, mutatorConfig.APIBackendHost, mutatorConfig.APIBackendPort, namespace, name)

	if mutatorConfig.Secure {
		client.Transport.(*http.Transport).TLSClientConfig = &tlsConfig
	}
	subs := entity
	if subscription != "" {
		subs = subscription
	}
	annotation := make(map[string]string)
	annotation[mutatorConfig.RemediationEventAnnotation] = alias
	check := &v2.Check{
		Subscriptions: []string{subs},
		Command:       command,
		Interval:      uint32(interval),
		Publish:       publish,
		Timeout:       uint32(10),
		RuntimeAssets: assets,
		Handlers:      handlers,
		ObjectMeta: v2.ObjectMeta{
			Name:      name,
			Namespace: namespace,
			Labels: map[string]string{
				mutatorConfig.Name: "owner",
			},
			Annotations: annotation,
			CreatedBy:   mutatorConfig.Name,
		},
	}
	if proxyEntity != "" {
		check.ProxyEntityName = proxyEntity
	}
	// s, err := json.MarshalIndent(check, "", "\t")
	// fmt.Println(string(s), url)
	encoded, _ := json.Marshal(check)
	req, err := http.NewRequest(http.MethodPut, url, bytes.NewBuffer(encoded))
	if err != nil {
		return fmt.Errorf("Failed to put event to %s failed: %v", url, err)
	}
	if len(mutatorConfig.APIBackendKey) == 0 {
		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", auth.AccessToken))
	} else {
		req.Header.Set("Authorization", fmt.Sprintf("Key %s", mutatorConfig.APIBackendKey))
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("error executing PUT request for %s: %v", url, err)
	}
	if resp.StatusCode < http.StatusOK || resp.StatusCode >= http.StatusMultipleChoices {
		return fmt.Errorf("PUT of event to %s failed with status %v\nevent: %s", url, resp.Status, string(encoded))
	}

	defer resp.Body.Close()

	return err
}

// used to clean errors output
func trimBody(body []byte, maxlen int) string {
	if len(string(body)) < maxlen {
		maxlen = len(string(body))
	}

	return string(body)[0:maxlen]
}
